File: src/main/java/top/toobee/optimization/cache/WardenCache.java
package top.toobee.optimization.cache;

import net.minecraft.core.BlockPos;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.entity.ai.memory.MemoryModuleType;
import net.minecraft.world.entity.monster.warden.Warden;
import net.minecraft.world.level.Level;
import org.apache.commons.lang3.tuple.Pair;

public final class WardenCache extends StackingMobCache<Warden> {
    public static final Caches<Warden, WardenCache> CACHES = new Caches<>(Warden.class, WardenCache::new);

    public int angerAtTarget = 0;

    private WardenCache(Level level, BlockPos pos) {
        super(level, pos);
    }

    @Override
    public void truncate() {
        CACHES.all.remove(Pair.of(level, pos));
    }

    @Override
    public void newSense(ServerLevel serverLevel, Warden entity) {
        senseNearestLivingEntities(serverLevel, entity);
        final var it = getNearestTarget();
        if (it == null)
            entity.getBrain().eraseMemory(MemoryModuleType.NEAREST_ATTACKABLE);
        else
            entity.getBrain().setMemory(MemoryModuleType.NEAREST_ATTACKABLE, it);
    }
}

File: src/main/kotlin/top/toobee/optimization/cache/WardenCache.kt
package top.toobee.optimization.cache

import net.minecraft.entity.ai.brain.MemoryModuleType
import net.minecraft.entity.mob.WardenEntity
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World

class WardenCache private constructor(
    world: World,
    pos: BlockPos,
) : StackingMobCache<WardenEntity>(world, pos) {
    companion object : Caches<WardenEntity, WardenCache>(WardenEntity::class.java) {
        override fun create(world: World, pos: BlockPos): WardenCache = WardenCache(world, pos)
    }

    var angerAtTarget: Int = 0

    override fun truncate() {
        super.truncate()
        all.remove(this.world to this.pos)
    }

    override fun newSense(world: ServerWorld, entity: WardenEntity) {
        this.senseNearestLivingEntities(world, entity)
        this.nearestTarget?.let {
            entity.brain.remember(MemoryModuleType.NEAREST_ATTACKABLE, it)
        } ?: entity.brain.forget(MemoryModuleType.NEAREST_ATTACKABLE)
    }
}